cdap-runtime-ext-emr/src/main/java/co/cask/cdap/runtime/spi/provisioner/emr/ElasticMapReduceProvisioner.java

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.runtime.spi.provisioner.emr;
+
+import co.cask.cdap.runtime.spi.provisioner.Cluster;
+import co.cask.cdap.runtime.spi.provisioner.ClusterStatus;
+import co.cask.cdap.runtime.spi.provisioner.Node;
+import co.cask.cdap.runtime.spi.provisioner.PollingStrategies;
+import co.cask.cdap.runtime.spi.provisioner.PollingStrategy;
+import co.cask.cdap.runtime.spi.provisioner.ProgramRun;
+import co.cask.cdap.runtime.spi.provisioner.Provisioner;
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerContext;
+import co.cask.cdap.runtime.spi.provisioner.ProvisionerSpecification;
+import co.cask.cdap.runtime.spi.ssh.SSHKeyPair;
+import co.cask.cdap.runtime.spi.ssh.SSHSession;
+import com.amazonaws.services.elasticmapreduce.model.ClusterSummary;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Provisions a cluster using Amazon EMR.
+ */
+public class ElasticMapReduceProvisioner implements Provisioner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ElasticMapReduceProvisioner.class);
+  private static final ProvisionerSpecification SPEC = new ProvisionerSpecification(
+    "aws-emr", "Amazon EMR",
+    "Amazon EMR provides a managed Hadoop framework that makes it easy, fast, and cost-effective to " +
+            "process vast amounts of data across dynamically scalable Amazon EC2 instances.",
+    new HashMap<>());
+  private static final String CLUSTER_PREFIX = "cdap-";
+
+  @Override
+  public ProvisionerSpecification getSpec() {
+    return SPEC;
+  }
+
+  @Override
+  public void validateProperties(Map<String, String> properties) {
+    EMRConf.fromProperties(properties);
+  }
+
+  @Override
+  public Cluster createCluster(ProvisionerContext context) throws Exception {
+    // Generates and set the ssh key
+    SSHKeyPair sshKeyPair = context.getSSHContext().generate("ec2-user"); // or 'hadoop'
+    context.getSSHContext().setSSHKeyPair(sshKeyPair);
+
+    EMRConf conf = EMRConf.fromProvisionerContext(context);
+    String clusterName = getClusterName(context.getProgramRun());
+
+    try (EMRClient client = EMRClient.fromConf(conf)) {
+      // if it already exists, it means this is a retry. We can skip actually making the request
+      Optional<ClusterSummary> existing = client.getUnterminatedClusterByName(clusterName);
+      if (existing.isPresent()) {
+        return client.getCluster(existing.get().getId()).get();
+      }
+      String clusterId = client.createCluster(clusterName);
```

Comment:
when we create the cluster, we need to make sure we're using the right emr version based on the spark compat value from ProvisionerContext.